src/Journal.cpp

```diff
@@ -145,6 +154,310 @@ wxWindow *FindByPath( const Path &path )
+
+   // Helper to keep casts out of the supplied function's definition
+   // Tag is the subtype of wxEvent expected by fn
+   template< typename Tag, typename SerialFn >
+   Serializer makeSerializer( SerialFn fn )
+   {
+      // Return an adaptor taking wxEvent
+      return [fn = std::move( fn )]( const wxEvent &e ){
+         return fn(static_cast<const Tag&>(e));
+      };
+   }
+
+   // Helper to check consistency
+   // Tag is a subtype of wxEvent (such as wxCommandEvent), and fn defines
+   // a deserializer dependent on a run-time wxEvent type (such as one of
+   // the many event types implemented by wxCommandEvent)
+   template< typename Tag, typename Fn >
+   Deserializer checkDeserializer( wxEventTypeTag<Tag> type, Fn fn )
+   {
+      // Check consistency of the deduced template parameters
+      // The deserializer should produce unique pointer to Tag
+      using DeserializerResult =
+         decltype( *fn( wxEVT_NULL, wxArrayStringEx{} ) );
+      static_assert( std::is_same_v< Tag&, DeserializerResult >,
+         "deserializer produces the wrong type" );
+
+      return std::move( fn );
+   }
+
+   // Type-erasing constructor applies compile-time type consistency checks to
+   // the functions, which can be supplied as lambdas using a subtype of wxEvent
+   template<
+      typename Tag, // such as wxCommandEvent
+      typename SerialFn,
+      typename DeserialFn
+   >
+   Type(
+      wxEventTypeTag<Tag> type, /*!< such as wxEVT_BUTTON
+         Tag is deduced at compile time, the int value is fixed at runtime
+      */
+      const Code &code, SerialFn serialFn, DeserialFn deserialFn )
+      : type{ type }
+      , code{ code }
+      , serializer{ makeSerializer<Tag>( std::move( serialFn ) ) }
+      , deserializer{ checkDeserializer( type, std::move( deserialFn ) ) }
+   {
+   }
+
+   // Data members
+   wxEventType type; //!< Just an int!
+   Code code; //!< Persistent cross-platform name corresponding to the int
+   Serializer serializer;
+   Deserializer deserializer;
+};
+
+wxString WindowEventName( const wxEvent &event )
+{
+   wxString windowName;
+   if ( auto pWindow = dynamic_cast< wxWindow * >( event.GetEventObject() ) )
+      windowName = Windows::FindPath( *pWindow ).GET();
+   return windowName;
+}
+
+std::optional<wxArrayStringEx> WindowEventSerialization( const wxEvent &event )
+{
+   std::optional< wxArrayStringEx > result;
+   if ( auto windowName = WindowEventName( event ); !windowName.empty() )
+      result.emplace( wxArrayStringEx{ windowName } );
+   return result;
+}
+
+template<typename Event = wxCommandEvent, typename EventType >
+static Type NullaryCommandType( EventType type, const wxString &code ){
+   const auto deserialize =
+   []( wxEventType evtType, const wxArrayStringEx &components) {
+      std::unique_ptr< wxCommandEvent > result;
+      if ( components.size() == 1 ) {
+         if ( auto pWindow = Windows::FindByPath( components[0] ) ) {
```

Comment:
Is that possible that for some reason this checks fail? What would be the consequence for the user?

Reply:
If the checks fail, then DispatchEvent will return false.  That causes Journal::Dispatch() to throw SyncException.   That exception's delayed handler action makes the program quit.  That causes Journal::GetExitCode() to return a non-zero status on the command line.

All of that is intended behavior, for the user who is the QA engineer.  All of this means an interaction sequence recorded before, fails to rerun now, because the sequence of events fails on rerun -- the windows that should be the handlers of events don't exist in the rerun.

Tests need some way to fail when they should fail and report that .

Reply:
Yes, if more details on failure will be ever needed that may be corrected in the future

Reply:
I tried to make the program return the line number of the journal file where failure occurred.  However at least on macOS it is truncated to an 8 bit value.